### DISCUSSION TITLE
Email verification for Guest Mode

### DISCUSSION THREAD

**Discussion by Lakshan-Madushanka** [Lakshan-Madushanka/laravel-comments#7] [Ideas]
I'm planning to implement email verification for guest mode. The current plan is as follows:

- Add an option named `must_verify_email` in the config file to toggle email verification.
- Move all guest-related data to a new `comment_guests` table.
- If the option is enabled, guests must verify their email to interact with comments (create, react, etc.).
- Send a verification link to the guest's email address.
- When the guest clicks the verification link, set the `email_verified` column to true along with the relevant `ip_address`.
- Configure the verification timeout in the config file.
- Use `verified_email + ip_address ` to authenticate a guest. This means the verified email is only valid for the relevant IP address. If the IP address differed, the guest must verify the email again to interact with comments associated with that email.
- I believe this is a robust way to authenticate users in guest mode. Without email verification, guests are authenticated solely by their IP address, which means the same user cannot interact with their own comments when using different devices with different IP addresses.

What are your thoughts?

**Comment by AkioSarkiz** [comment]
Hey, 

In general it sounds nice, I have a small suggestion:

Verification email is equal to registration In most situations. So it makes sense to delegate logic of creation not verified user to an application or application can create verified user if the application doesn't require email verification for creating a new account. Then you can use just Auth facade to check permission

**Reply by Lakshan-Madushanka** [reply]
@AkioSarkiz Thanks for the feedback. Could you elaborate on this? Some points are a bit subtle to comprehend.

**Reply by AkioSarkiz** [reply]
The idea is to make package doesn't responsible of auth user. We have auth system in the laravel, so the package can provide interface like `VerificableComment` that user should implement to the package can auth user's actions like create/update/delete.

The config file may look like:

```php
[
....
    /**
     * Comment owner model
     * Must extend Illuminate\Contracts\Auth\Authenticatable
     * Must implement LakM\Comments\Contracts\CommentableContract
     * Optional: Should implement VerificableComment to enable gust comments verification
     */
    'user_model' => User::class,
....
]

**Reply by Lakshan-Madushanka** [reply]
Actually, in auth mode, it works that way.The package doesn't involve any auth-related functionality. But in guest mode, we have to find alternative ways to uniquely identify users, as in guest mode we assume no auth functionality is available, and the package doesn't need to implement them as it is completely out of the scope.

**Reply by AkioSarkiz** [reply]
This is a difference of our opinions. I think that it is better to register users as a model to identify them. On the other hand, you think the package should make it in an alternative way.

Let's see on the cons and pros of the first method.


Pros:
- User hasn't to auth each time when he wants to modify or create a comment
- More securely, because to log in as a user we need email and password
- More flexibility, in my opinion. For example, if my website does not require email verification, then I can just create a user which a want to.

Cons:
- More complicated
- Require more resources for implementation

**Reply by Lakshan-Madushanka** [reply]
That's another way, but it could be a bad idea to mess with the user auth model. For example, imagine a website that has an auth system only for admin users and uses the guest mode of the package. If you use the auth system for package users to log in too, then they can also log in to the admin panel. So, we have to keep different auth providers, models, and tables for the package. Isn't that messier than the first approach?

**Reply by AkioSarkiz** [reply]
I see a simple way to resolve the issue that you describe. This application can create a custom guest model instead of using user model and then make a custom auth guard for that guest model. Then everything should be fine. This is flexibility that I mention above

**Reply by Lakshan-Madushanka** [reply]
Thanks. I will consider this.

**Comment by Lakshan-Madushanka** [comment]
Implemented in https://github.com/Lakshan-Madushanka/laravel-comments/releases/tag/2.0.0
